### PR TITLE
Support opening locations in new window

### PR DESCRIPTION
**Pull request by ronmrdechai**
This patch adds support for opening locations (go to definition, declaration, etc.) in a new window (split) vs the current window or a preview window. I couldn't find a way to ask for this explicitly. This patch adds new commands prefixed with `:Split` or `-split` to support this.

### CODE REVIEW

**Comment by hrsh7th** [comment]
I think it can be better to change to `LspDefinition split` instead of `LspSplitDefinition`. (`edit` as the default value is good)

This IMO makes possible to use `tabedit`, `vsplit` or `split`.

What do you think about it?

**Comment by ronmrdechai** [comment]
To add an argument for `:LspDefinition` ? How does that enable `tabedit`, `vsplit` or `split`.

**Comment by thomasfaingnaert** [comment]
Instead of adding an `in_split` argument, maybe we can repurpose the `in_preview` to be a more general "where to open the location" parameter.
That way, we can support preview, popup, split, vsplit, tabs, ... without adding a zillion different parameters.

**Comment by hrsh7th** [comment]
You can use the following codes.

```viml
command! -nargs=1 LspDefinition call lsp#ui#vim#definition(0, '<args>')
```

(Originally, you need to add `-complete=customlist,s:edit_commands`, but we can improve after this PR)

**Comment by hrsh7th** [comment]
Agree, I think we change to `edit_cmd` instead of `in_preview` is better (`preview` as a special type in `edit_cmd`)

**Comment by ronmrdechai** [comment]
I’ll get rid of `in_preview` along the way

**Comment by ronmrdechai** [comment]
Updated PR. `in_preview` as been replaced with a `where` parameter. Not sure what `edit_cmd` was so I just went with this. `:LspDefinition` now supports `split`, `vsplit` and `preview` arguments.

File: plugin/lsp.vim [hunk 1]
```diff
@@ -80,30 +80,30 @@ command! LspNextDiagnostic call lsp#ui#vim#diagnostics#next_diagnostic()
...
+command! -bang -nargs=? -complete=customlist,lsp#ui#vim#open_where LspImplementation call lsp#ui#vim#implementation('<args>')
+command! LspPeekImplementation call lsp#ui#vim#implementation('preview')
 command! -nargs=0 LspStatus call lsp#print_server_status()
 command! LspNextReference call lsp#ui#vim#references#jump(+1)
 command! LspPreviousReference call lsp#ui#vim#references#jump(-1)
 command! -nargs=? -complete=customlist,lsp#server_complete LspStopServer call lsp#ui#vim#stop_server(<f-args>)
 command! -nargs=? -complete=customlist,lsp#utils#empty_complete LspSignatureHelp call lsp#ui#vim#signature_help#get_signature_help_under_cursor()
-command! LspDocumentFold call lsp#ui#vim#folding#fold(0)
-command! LspDocumentFoldSync call lsp#ui#vim#folding#fold(1)
+command! LspDocumentFold call lsp#ui#vim#folding#fold('default')
```
Comment (thomasfaingnaert) [inline plugin/lsp.vim]: The parameter refers to sync/async, so you shouldn't change this one :smile:

Comment (ronmrdechai) [reply plugin/lsp.vim]: Oops. Did this with regex 😛

File: plugin/lsp.vim [hunk 2]
```diff
@@ -65,10 +65,10 @@ command! -range -nargs=* -complete=customlist,lsp#ui#vim#code_action#complete Ls
       \   'selection': <range> != 0,
       \   'query': '<args>'
       \ })
-command! LspDeclaration call lsp#ui#vim#declaration(0)
-command! LspPeekDeclaration call lsp#ui#vim#declaration(1)
-command! LspDefinition call lsp#ui#vim#definition(0)
-command! LspPeekDefinition call lsp#ui#vim#definition(1)
+command! -bang -nargs=? -complete=customlist,lsp#ui#vim#open_where LspDeclaration call lsp#ui#vim#declaration('<args>')
+command! LspPeekDeclaration call lsp#ui#vim#declaration('preview')
```
Comment (ronmrdechai) [inline plugin/lsp.vim]: Peek commands are kept around for backwards compatibility. They are identical to `:Lsp<Command> preview`.

**Comment by ronmrdechai** [comment]
@thomasfaingnaert can this be merged or is more work required?

**Comment by thomasfaingnaert** [comment]
Looks good, just needs an update to the documentation.
There's also an alternative for this at #776, which uses modifiers instead.
I think I slightly prefer that, since it uses native Vim functionality.

@ronmrdechai What do you think about the other PR? Are you still missing anything with that implementation?

@hrsh7th Do you have a preference w.r.t. #776 vs #763?

**Comment by ronmrdechai** [comment]
Honestly #776 was what I was going for originally, but I couldn't figure out how to pass the `<mods>` in. I'd merge that instead. Keeping PR this open until a decision is made.

**Comment by prabirshrestha** [comment]
`<mods>` PR has been merged. Please use it instead. https://github.com/prabirshrestha/vim-lsp/pull/776

Thanks for the PR.

**Pull request closed without merging**